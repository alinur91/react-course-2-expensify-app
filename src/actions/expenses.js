import uuid from 'uuid';

// ADD_EXPENSE
export const addExpense = (
    {
        description='',note='', /* default values */
        amount=0,
        createdAt=0
    }={}
) => ({ 
/* if object doesnt exist we will desctrutre empty object */
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

//REMOVE_EXPENSE
export const removeExpense = ({id}={}) =>( /* {id}={} degen zna4it dispatch(removeExpense({id})) degen soz object peredaem */
    {
        type: 'REMOVE_EXPENSE',
        id
    }
)

//EDIT_EXPENSE
export const editExpense = (id,updates) =>(
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
)