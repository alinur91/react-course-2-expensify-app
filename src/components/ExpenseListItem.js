import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral'

/* kogda componenty podklu4aesh connect() k store {expenses: [{}],filters:{}} to 4erez propsa est dostyp u componenty k dispatch(someAction()) */
 const ExpenseListItem = ({id,description,amount,createdAt}) => ( /* sdez props desctrutre.<ExpenseListItem key={expense.id} description={expense.description},amount:{expense.amount},createdAt: {expense.createdAt} /> */
    <div>
    <Link to={`/edit/${id}`}> <h3 > {description}. </h3> </Link> 
        <p>
        {numeral(amount/100).format('$0,0.00')} 
        - 
        {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
       
    </div>
) 
//ExpenseListItem it takes some data in it renders smth,add a shapshot for what it renders.Its going to allow us to track changes to the components over time
export default ExpenseListItem /* v connect() v function ne zasovivaem potomy shto nam state ne nuzhen {expenses: [{}],filters:{}},mi connectiem shtoby k dispatch dostyp bil u componenty */
/*  dispatch(removeExpense({id})) delaem shtoby combineStore(expenses: ExpenseReducer) zarobatal vernet updated array, i s pomoshu connecta berem stae i renderem v componente ExpensesList renderim */

/* Kogda peredaem v instance ExpenseListItem propsi <ExpenseListItem  key={expense.id} description={expense.description} amount={expense.amount} createdAt={expense.createdAt} />  to mozhno destructure v samom componente*/

/*When we use connect you do not take anything from the state,we just neeed disptach,we dont need anthting from the state you just want to connect it to be able to access dispatch  */