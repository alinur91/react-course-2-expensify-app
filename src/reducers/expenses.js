// Expenses Reducer
const expensesReducerDefaultState = []

export default  (state=expensesReducerDefaultState ,action) =>{
    // console.log(state)
/* return a new state? */
switch (action.type) { /* push changes original array,we don't want to change state or action,we just want to read off of it */
case 'ADD_EXPENSE': return [...state,action.expense]//state degen: [{id:123,description:'asd',note:'',amount:123,createdAt: 456}] action.expense degen {id:123,description:'asd',note:'',amount:123,createdAt: 456} /* returns a new array,return array soxranitsya k expensesReducer variablu */
/* aside from concat we can use spread operator.action degen object vnutri property expense,id,amount */
    case 'REMOVE_EXPENSE': return state.filter(({id})=> id !==action.id)
    case 'EDIT_EXPENSE': return state.map(expense=>{
/* action degen   {type: 'EDIT_EXPENSE',id,updates} updates degen {amount: 900} */
/* state degen [{id: "bed68040-5eed-4835-81d9-1ec45e9f06bf", description: "Rent", note: "rent for may", amount: 800, createdAt},{id: "9f0a4e1c-1f5c-4085-a425-277adb9116e8", description: "Buy a car", note: "", amount: 10000, createdAt: 0}] */
        if(expense.id === action.id){
            return { /* i want to return a new object,changing expense,i want to grab all existing properties,and overwrite any of the ones that were passed down */
                ...expense,
                ...action.updates
            }
        }else{
            return expense
        }
    })
    default: return state    
}
}

