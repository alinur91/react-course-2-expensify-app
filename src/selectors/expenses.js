import moment from 'moment';

// Get visible expenses
export default (expenses,{text,sortBy,startDate,endDate}) => { /* vse dispatch sdelali neskolko zdes polnaya informacia,s pomoshui etogo vernem array [{},{}] */
    return expenses.filter(expense=>{
const createdAtMoment = moment(expense.createdAt) //expense.createdAt cifra goi
const startDateMatch= startDate? startDate.isSameOrBefore(createdAtMoment,'day') : true; //if there is no start date we're never going to filter things out based off of that,startDate always be true if there is no startdate
//typeof startDate !== 'number' || expense.createdAt >= startDate; /*only a start date is a number do we want to filter expenses out  */
/*lets say startdate is undefined,and createdAt is 1,it will stay(wont be filtered).Lets say startDate is 2 and created_at is 1(it will not stay,sozdannyi data ranshe bil,a user pozhe pokazivaet startDate) etot false budet startDateMatch  */
const endDateMatch= endDate? endDate.isSameOrAfter(createdAtMoment,'day') : true;
//typeof endDate !== 'number' || expense.createdAt <= endDate;
const textMatch=expense.description.toLowerCase().includes(text.toLowerCase())

        
/* esli vse eto startDateMatch && endDateMatch && textMatch bydet true to  expense ostaetsya v arraye*/
        return startDateMatch && endDateMatch && textMatch
    }).sort((a,b)=>{ /* a,b degen object {id: 123,description:'asd',createdAt: 123,amount:123} */
        if(sortBy === 'date') { /* sortiruem po recent data */
            return a.createdAt > b.createdAt?-1:1
        } // I want to see my most expenses coming 1st,from sorting date i want to see recent expenses coming 1st
        else if(sortBy === 'amount'){
            return b.amount - a.amount /* [1(a),2(b)] b-a = polozhitelnoe chislo,zna4it menyaem mestami,sonda kemy retimen boladi 800,100,50 */
           // a.amount > b.amount?-1:1
        }
    })
}

//JEST testing framework