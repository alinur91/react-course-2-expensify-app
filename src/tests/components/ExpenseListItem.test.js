import React from 'react'; /* importiruem React potomy shto componenty <ExpenseListItem/> ispolzuem */
import ExpenseListItem from '../../components/ExpenseListItem'; 
import {shallow} from 'enzyme'; // import because we're going to be using that to render our component
import expenses from '../fixtures/expenses'; /* array vidumanniy koldanamyz */

//ExpenseListItem it takes some data in it renders smth,add a shapshot for what it renders.Its going to allow us to track changes to the components over time

test('should render ExpenseListItem with expense',()=>{ // It expects few props {id,description,amount,createdAt}
const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>) /*{...expenses[0]} degen ===> <ExpenseListItem id='123' description='rent' amount=900 />  */
/*this adds object expenses[0], this ... spreads it out adding all of its properties as props to ExpenseListItem  */
/* <ExpenseListItem/> vizivaem ety componenty i polu4aem JSX==><div>
    <Link to={`/edit/${id}`}> <h3 > {description}. </h3> </Link> 
        <p>{amount} - {createdAt}</p>
    </div> */
    expect(wrapper).toMatchSnapshot()  /* make assertion */
/* esli naprimer v ExpenseListItem.js 4eto kakoi-nibud symbol dobavim ne4aino,zdez test will notify it */

})