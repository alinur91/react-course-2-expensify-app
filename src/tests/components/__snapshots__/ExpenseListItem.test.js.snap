// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/123"
  >
     
    <h3>
       
      xbox
      . 
    </h3>
     
  </Link>
  <p>
    1120
     - 
    0
  </p>
</div>
`;
